internal/querycoord/impl.go

```diff
@@ -88,6 +88,58 @@ func (qc *QueryCoord) GetStatisticsChannel(ctx context.Context) (*milvuspb.Strin
 	}, nil
 }
 
+func (qc *QueryCoord) checkAnyReplicaAvailable(collectionID UniqueID) bool {
+	replicas, _ := qc.meta.getReplicasByCollectionID(collectionID)
+	// Note: copy from GetShardLeaders
+	shards := make(map[string]*querypb.ShardLeadersList)
+	shardNodes := getShardNodes(collectionID, qc.meta)
+	for _, replica := range replicas {
+		for _, shard := range replica.ShardReplicas {
+			list, ok := shards[shard.DmChannelName]
+			if !ok {
+				list = &querypb.ShardLeadersList{
+					ChannelName: shard.DmChannelName,
+					NodeIds:     make([]int64, 0),
+					NodeAddrs:   make([]string, 0),
+				}
+			}
+
+			isShardAvailable, err := qc.cluster.IsOnline(shard.LeaderID)
+			if err != nil || !isShardAvailable {
+				log.Warn("shard leader is unavailable",
+					zap.Int64("collectionID", replica.CollectionID),
+					zap.Int64("replicaID", replica.ReplicaID),
+					zap.String("DmChannel", shard.DmChannelName),
+					zap.Int64("shardLeaderID", shard.LeaderID),
+					zap.Error(err))
+				continue
+			}
+
+			nodes := shardNodes[shard.DmChannelName]
+			for _, nodeID := range replica.NodeIds {
+				if _, ok := nodes[nodeID]; ok {
+					if ok, err := qc.cluster.IsOnline(nodeID); err != nil || !ok {
+						isShardAvailable = false
+						break
+					}
+				}
+			}
+
+			if isShardAvailable {
+				list.NodeIds = append(list.NodeIds, shard.LeaderID)
+				list.NodeAddrs = append(list.NodeAddrs, shard.LeaderAddr)
+				shards[shard.DmChannelName] = list
+			}
```

Comment:
As previously discussed, this part checks whether one shard can provided service. What we want shall be whether each shard has at least one replica could provide query service.

Reply:
Gonna fix it later.

Reply:
Done